### PR TITLE
Report system feature implementation

### PR DESCRIPTION
**Pull request by bulgit**
Following the issue #52, I decided implement the report feature! I think it will be a lot beneficial for everyone in the Microsoft Community (moderators & users).

### Commands

- `!report_fix` is called `!reportf` instead, and anyone can fix a report as long as they own it (or if it's a moderator).
- `!report`, `!reportf`, and `!reports` can be processed in DM as well so that no one would be able to know who reported them.
- Added `!report_u`: allows anyone to report an user from the server. This should only be used if the user is doing things outside the server, like sending malicious DMs.
- Added `!reports`: allows anyone to view their report statistics.

### Reports

There are currently 2 report types: **Message**, and **User**.

Whenever an user reports something, the bot will generate a report message in a specific channel **#reports** channel (must be accessible by mods only), configurable in _config.json_. The report is more detailed, it includes the type, reason, report ID, status, list of people who reported and the moderator who reviewed the report.
Also, to alert moderators the bot will ping **@here** for every generated pending report message in the **#reports** channel (configurable in _config.json_).

#### Interaction

Now there are currently 4 buttons, for handling the report:
- Warn and clean => if the reported content is a message, the message will be deleted. If it's an user, the user will only be warned (same as _Warn only_)
- Warn only
- No warn
- Decline report
Reactions emoji are configurable in _config.json_.

Once the report was reviewed, the generated report message is moved to the **#reviewed-reports** channel. It is better for moderators, reports are split between pending reports and reviewed reports.
"decline" is called "reject" in the code, maybe decline is more clever/satisfying to the user, not sure which word is better. Same for "accept", this one may be the most suitable word for reports.

### More new conditions

- If a moderator deletes the generated pending report message (from **#reports** channel), the bot will recreate it.
- On initialization, the bot will process pending reports. This includes checking for interactions (moderator who reacted to the generated pending report message), and checking for generated pending report messages that were deleted (will recreate them in this case).

The following conditions are applied only in non-Debug builds:
- A moderator can't be reported (configurable in _config.json_).
- The bot can't be reported.
- Users can't report themselves.

#### Reported messages

- A message that was reviewed can't be reported again, except if it was edited (prevent abuse/exploit).
- A generated report message contains the reported content at the time of the report. The report won't be edited if the original reported message is edited. This should prevent trolls from avoiding punishment if they see that they were reported.

### Notes

The report system considers users as moderator if they have access to **#reports** channels. Maybe make a list of roles that cannot be reported instead? Also, my suggestion is to avoid writing anything inside any of these **#reports** channel. It should be better to discuss about a report in another private channel instead (I've seen that there is an **#investigation** channel, could be there!).

This feature may require more testing. Also generated texts in _Modules/Reports.cs_ could be done better (I'm not a native english speaker).
The only thing I'm worried about is race-condition, but I think that the bot waits for the async command to finish, it should be good.

There is 1 issue: currently, generated report messages can't contain uploaded images. This could be fixed by replacing the reported message image with the discord link to the uploaded image.

### What has been tested

- [x] Report messages
- [x] Report users
- [x] Boost a report (one or more users reporting the same content)
- [x] Make multiple reports
- [x] Fixing a report that was reviewed
- [x] Fixing an owning report
- [x] Fixing a report as a moderator
- [ ] Fixing a report without owning it and without being a moderator
- [x] Fixing a report, and review the report
- [x] View report stats
- [x] Accept a report (warn & clean, warn, no warn)
- [x] Reject a report
- [x] Delete a generated pending report message
- [x] Shutdown the bot, delete a generated pending report message and restart the bot
- [x] Shutdown the bot, react to a generated pending report message and restart the bot
- [x] Delete/edit the original reported message, and then review the report
- [x] Review a report, then edit the reported message and report the message again
- [ ] Docker container

### CODE REVIEW

File: Cliptok.csproj [hunk 1]
```diff
@@ -10,10 +10,10 @@
 
 	<ItemGroup>
 		<PackageReference Include="Abyssal.HumanDateParser" Version="2.0.0-20191113.1" />
-		<PackageReference Include="DSharpPlus" Version="4.1.0-nightly-00892" />
-		<PackageReference Include="DSharpPlus.CommandsNext" Version="4.1.0-nightly-00892" />
+		<PackageReference Include="DSharpPlus" Version="4.0.1" />
+		<PackageReference Include="DSharpPlus.CommandsNext" Version="4.0.1" />
```
Comment (Erisa) [inline Cliptok.csproj:14]: We were on nightly for a reason, please don't do this.

**Comment by Erisa** [comment]
Honestly I'm stunned that you put in the time and energy to write 2000 lines of code that might never be used because the feature you suggested hasn't even been discussed internally, nevermind approved.

Personally I don't think a `!report` system is necessary when we already have a perfectly functional system with modmail that users have been happy with for years. If i did add a way to quickly report something, it would just open a modmail thread because thats our preferred way to discuss and take action together as a team. Not every situation is clear cut and sometimes mods don't know what to do, so a simple `#reports` message doesn't quite cut it.

It puts me in a difficult position because I don't want to merge a feature that we don't want to use because of one reason or another, but if I reject this request then it looks bad on me for throwing away alll your hard work.
Generally we prefer if feedback is presented by modmail (An Issue is less noticed, I was going to think about it later but it slipped my mind in the hurry of other thing) because from our point of view it's significantly easier to manage and respond to.

File: Program.cs [hunk 2]
```diff
@@ -299,14 +299,15 @@ async Task MessageUpdated(DiscordClient client, MessageUpdateEventArgs e)
...
-            }); ;
+            });
 
             commands.RegisterCommands<Warnings>();
             commands.RegisterCommands<MuteCmds>();
             commands.RegisterCommands<UserRoleCmds>();
             commands.RegisterCommands<ModCmds>();
             commands.RegisterCommands<Lockdown>();
             commands.RegisterCommands<Bans>();
+            commands.RegisterCommands<Reports>();
```
Comment (superkid200) [inline Program.cs:310]: Should the !report command added as a slash command for more visibility?

**Review by superkid200 — CHANGES_REQUESTED** [review]

**Comment by bulgit** [comment]
~2000 lines not really much (for me, keeps myself busy and it's challenging), I may reuse the code/design to make a custom bot. I didn't know anyone could talk to mod mail about things like this because I had been told (by a mod) that every requests / issue about the bot should be made here on GitHub.
Btw I'm not expecting anything in return, it's fine. I just offer some help the Microsoft Community since it's a great place!

About Mod mail, it's a nice system and I don't dislike it (not looking for removing it at all). It's just that the report system focuses on one specific thing and may allow mods/users do reports faster. There is 1 thing I could agree with is that the reviewer won't be able to tell the user why the report was rejected (users can still DM mod mail tho), but it still could be fixed with a new command.
Users may use this feature if they are aware about it (as in everywhere).

I understand that this feature may cause reluctance especially if it adds more confusion for moderators. Honestly I don't know what to say, I still think this feature is useful because it's basically a one-click-review for moderators (if you are really busy it could be useful). Maybe others can tell what they think about it?

File: Program.cs (hunk 2 above)
Comment (bulgit) [reply Program.cs:310]: I think it would be a good idea to make users aware about the command (only /report and /report_u maybe?)

**Comment by Erisa** [comment]
> I didn't know anyone could talk to mod mail about things like this because I had been told (by a mod) that every requests / issue about the bot should be made here on GitHub.
Well it is true that you can suggest ideas for the bot on GitHub and we did notice, it just seems like the kind of major change to the way we approach moderation that would require input from those not watching this repository (Only a handful of us actually check here, and the bot is basically my hobby project more than anything). As opposed to a general improvement to the bot that might not affect the entire moderation flow.
Generally suggestions for the server and the way it operates itself is done through modmail. I can understand the confusion though, and it isn't an issue really we just hadn't got around to discussing the issue and only a couple of us noticed. Modmail has better visibility if you want better opinions.

> Honestly I don't know what to say, I still think this feature is useful because it's basically a one-click-review for moderators (if you are really busy it could be useful). Maybe others can tell what they think about it?
That's a good idea. For starters I'm curious if @torchgm, @FloatingMilkshake or @TheTimeWalker has anything to say about this suggestion. No pressure to them though.

**Comment by torchgm** [comment]
Personally I don't see much benefit in such a system, it'll just lead to an increase of low-quality reports. We'll still have to check the context of a lot of the messages reported, and if it's a larger issue at play we don't get the benefit of a dedicated channel to discuss it in (something that modmail provides). Furthermore, modmail is nice because we can follow up and ask users to elaborate as to why they're reporting things, in the event the context isn't immediately clear.

The difference in time between messaging modmail (which can be done inline directly from the user profile without leaving the server) and typing out a command isn't very significant, nor is the time saved for mods by not having to type out a moderation command, especially with the assistance of slash commands.

Lastly, I can see there being a couple of moderation and safety-related issues regarding such an addition. Malicious users with custom clients or a keen eye could see who was reporting their messages and potentially target them; unfortunately humans suck and this is something that we can't take for-granted. This can be addressed simply by incorporating a slash command, however it would mainly seem to be a stumbling block and another layer of confusion for users. "Do I report something in modmail or with the report command?" would be a question I could imagine a lot of new users asking, which will inevitably lead to one of two responses in chat:

 - "What are you trying to report?"
 - "If you're not sure then just send it to modmail/use the report command"

The former would encourage discussion of rule-breaking activity, risk leading into a bunch of unnecessary drama and make it obvious who's snitching on who (again putting that user at risk of harassment). The latter would just prioritise one system over the other, either losing us the benefits of modmail or leaving the report command relatively unused.

Sorry to disappoint, and this is just my opinion, but I just don't really think it would be a particularly wise move to implement something like this right now.

File: Program.cs (hunk 2 above)
Comment (superkid200) [reply Program.cs:310]: I think so, just hide the mod commands like agree with complaint or ignore.

**Comment by superkid200** [comment]
Yeah, one concern is there might be people who will use this command to spam/overwhelm mods (server raid). There are 2 solutions I can think of:

- Only allowed users with a specific tier to use the command
- Set this command as a privilege, and will revoke access to those who abuse it.

**Comment by Erisa** [comment]
Hi, since we decided as a mod team to not use this feature and recieved no response and the files have changed a lot since, I am going to close this PR.

You are more than welcome to have a report system in your own bot or server, but we are not going to support it as part of Cliptok and Microsoft Community.

**Pull request closed without merging**